manifest/v1alpha/slo/validation.go

```diff
@@ -37,27 +44,96 @@ var specValidationNonComposite = validation.New[Spec](
 var specValidationComposite = validation.New[Spec](
 	validation.For(func(s Spec) Indicator { return s.Indicator }).
 		WithName("indicator").
+		StopOnError().
 		Rules(
 			validation.Forbidden[Indicator]().WithDetails(
 				"indicator section is forbidden when spec.objectives[0].composite is provided",
 			),
-		).
-		StopOnError(),
+		),
 
 	validation.ForPointer(func(s Spec) *Composite { return s.Composite }).
 		WithName("composite").
+		StopOnError().
 		Rules(
 			validation.Forbidden[Composite]().WithDetails(
 				"composite section is forbidden when spec.objectives[0].composite is provided",
 			),
-		).
-		StopOnError(),
+		),
 
 	validation.For(func(s Spec) Spec { return s }).
 		StopOnError().
 		Rules(specCompositeObjectiveValidationRule),
+
+	validation.Transform(func(s Spec) string { return s.Objectives[0].Composite.MaxDelay }, time.ParseDuration).
+		WithName("objectives[0].composite.maxDelay").
+		StopOnError().
+		Rules(
+			validation.DurationPrecision(time.Minute),
+			validation.GreaterThanOrEqualTo(time.Minute),
+		),
+
+	validation.For(func(s Spec) []CompositeObjective { return s.Objectives[0].Composite.Objectives }).
+		WithName("objectives[0].composite.components[0].objectives").
+		StopOnError().
+		Rules(
+			validation.SliceMinLength[[]CompositeObjective](2),
+		),
+
+	// Weights of composite objectives must be greater than zero
+	validation.ForEach(func(s Spec) []CompositeObjective { return s.Objectives[0].Composite.Objectives }).
+		WithName("objectives[0].composite.components[0].objectives").
+		StopOnError().
+		IncludeForEach(compositeObjectiveRule),
 ).When(func(s Spec) bool { return s.HasCompositeObjectives() })
 
+var sloValidationComposite = validation.New[SLO](
+	validation.For(func(s SLO) SLO { return s }).
+		StopOnError().
+		Rules(
+			validation.NewSingleRule(func(s SLO) error {
+				for _, obj := range s.Spec.Objectives[0].Composite.Objectives {
+					isSameProject := obj.Project == s.Metadata.Project
+					isSameName := obj.Objective == s.Metadata.Name
+
+					if isSameProject && isSameName {
+						return validation.NewPropertyError(
+							"slo",
+							s.Metadata.Name,
+							errors.Errorf("composite SLO cannot have itself as one of its objectives"),
+						)
+					}
+				}
+
+				return nil
+			}).WithErrorCode(validation.ErrorCodeForbidden),
+		),
+
+	validation.For(func(s SLO) []CompositeObjective { return s.Spec.Objectives[0].Composite.Objectives }).
+		StopOnError().
+		Rules(
+			validation.NewSingleRule(func(c []CompositeObjective) error {
+				sloMap := make(map[string]bool)
+
+				for objKey, obj := range c {
+					key := fmt.Sprintf("%s%s%s", obj.Project, obj.SLO, obj.Objective)
```

Comment:
Assuming we have the following two objectives:
```
- project: a
  slo: bc
  objective: d
- project: a
  slo: b
  objective: cd
```
They will concatenate into identical key `abcd` and trigger a validation error. Let's add a separator that is forbidden in project, slo, objective name:
```suggestion
					key := fmt.Sprintf("%s/%s/%s", obj.Project, obj.SLO, obj.Objective)
```

Reply:
Makes sense, changed